### PR TITLE
Hash linker

### LINKED ISSUE #4429: Cropped link path in generated bytecode
If you run:

`solc path-to-somefile/somefile.sol (other options)`, the generated bytecode, instead of doing a reference like this:

`__SafeMath.sol:SafeMath__`

inserts this cropped output:

`__src/contracts/SafeMath.sol:Safe__`

Regards,
Mauro.

### PR DESCRIPTION
**Pull request by chriseth**
Fixes #3918 
Fixes #4429

### CODE REVIEW

**Comment by chriseth** [comment]
It seems this needs an implementation of the linker in solc-js and for that we need a keccak256 dependency.

**Comment by chriseth** [comment]
Needs https://github.com/ethereum/solc-js/pull/266

**Comment by axic** [comment]
I think standard json output should contain the hash for the unlinked references or at least the documentation be extended to explain that.

**Comment by chriseth** [comment]
@axic standard json contains offset references - I don't think we need to list the hashes in addition to that.

**Comment by erak** [comment]
This looks very good. I'd consider some more testing around compiling and linking multiple libraries. Even though conflicts should not happen anymore in theory. I feel like having some kind of proof with regards to the audit.

@axic I also think that we do not need to add extra hashes for unlinked references to the standard json output. But I'd test the bytecode format more extensively and then cover the offsets.

**Review by erak — CHANGES_REQUESTED** [review]
After going through this again, I think that we should have enough test coverage. But: the documentation needs an update.

**Comment by chriseth** [comment]
Added documentation - I actually already had it, just forget to check it in.

File: docs/using-the-compiler.rst [hunk 1]
```diff
@@ -41,14 +41,25 @@ If there are multiple matches due to remappings, the one with the longest common
 
 For security reasons the compiler has restrictions what directories it can access. Paths (and their subdirectories) of source files specified on the commandline and paths defined by remappings are allowed for import statements, but everything else is rejected. Additional paths (and their subdirectories) can be allowed via the ``--allow-paths /sample/path,/another/sample/path`` switch.
 
-If your contracts use :ref:`libraries <libraries>`, you will notice that the bytecode contains substrings of the form ``__LibraryName______``. You can use ``solc`` as a linker meaning that it will insert the library addresses for you at those points:
+If your contracts use :ref:`libraries <libraries>`, you will notice that the bytecode contains substrings of the form ``__53aea86b7d70b31448b230b20ae141a537e5__``. These are placeholders for the actual library addresses.
+The placeholder is a 36 character prefix of the hex encoding of the keccak256 hash of the fully qualified library name.
+The bytecode file will also contain lines of the form ``// <placeholder> -> <library name>`` at the end to help
```
Comment (erak) [inline docs/using-the-compiler.rst]: Shouldn't this be ``// <placeholder> -> <library path>:<library name>``?

Comment (chriseth) [reply docs/using-the-compiler.rst]: I will change it to "fully qualified library name".

File: solc/CommandLineInterface.cpp [hunk 2]
```diff
@@ -905,8 +905,10 @@ void CommandLineInterface::handleCombinedJSON()
 		if (requests.count("metadata"))
 			contractData["metadata"] = m_compiler->metadata(contractName);
 		if (requests.count(g_strBinary))
+			// Do we want to change it here or not?
```
Comment (erak) [inline solc/CommandLineInterface.cpp]: What do you mean with this comment? Is this a left-over?

**Comment by erak** [comment]
Restarted the failing Linux test.

File: solc/CommandLineInterface.cpp (hunk 2 above)
Comment (chriseth) [reply solc/CommandLineInterface.cpp]: It is a question to the reviewers.

Comment (erak) [reply solc/CommandLineInterface.cpp]: Well, I wasn't sure :) I'd prefer to add such questions as regular comments to the pull request. In that case, I'd say yes, let's be consistent and also add the reference hints to the ``combined-json`` output.

Comment (chriseth) [reply solc/CommandLineInterface.cpp]: Ok, perhaps I should give the reason why I did not add it here: I would assume some tools to be confused when a json string suddenly starts containing newlines, and several tools might not be prepared for that.

**Comment by chriseth** [comment]
The driveby-bugfixes inside this PR are getting out of hand...

**Review by erak — DISMISSED** [review]

**Pull request merged**